### PR TITLE
Add solution removal agent

### PR DESCRIPTION
**Pull request by kilativ-dotcom**

### CODE REVIEW

**Review by MksmOrlov — CHANGES_REQUESTED** [review]

File: problem-solver/cxx/solutionModule/manager/DeleteSolutionManager.cpp [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+{
+  if (solution.IsValid() == SC_FALSE)
+    SC_THROW_EXCEPTION(utils::ExceptionItemNotFound, "DeleteSolutionManager: solution is not valid");
+  ScAddrList const & ruleAndSubstitutionPairs = getListFromSet(solution);
+  SC_LOG_DEBUG("DeleteSolutionManager: Solution has " << ruleAndSubstitutionPairs.size() << " elements");
+  safeEraseElement(solution);
+  deleteRuleAndSubstitutionsPairs(ruleAndSubstitutionPairs);
+}
+
+ScAddrList DeleteSolutionManager::getListFromSet(ScAddr const & set) const
```
[the remark this answers is not in the document]
Comment (kilativ-dotcom) [reply problem-solver/cxx/solutionModule/manager/DeleteSolutionManager.cpp:24]: looks like sc-machine is unable to iterate over Iterator3 while removing nodes that are at iterator->Get(2)

File: docs/scl-machine.tex [hunk 2]
```diff
@@ -333,3 +331,24 @@
...
+	принадлежащего классу \textit{действие. удалить решение}. После проверки sc-агентом условия инициирования
+	выполняется удаление дерева решения.}
+\scnrelfrom{пример входной конструкции}{\scnfileimage[45em]{images/delete_solution_tree_input.png}}
+\begin{scnrelfromvector}{аргументы агента}
+	\scnitem{\_solution\_tree}
+	\begin{scnindent}
+		\scnidtf{solutionTree}
+		\scnidtf{solution}
+		\scntext{пояснение}{Дерево решения, которое требуется удалить.}
+		%\scnrelfrom{описание примера}{\scnfileimage[28em]{images/target_template.png}}
```
Comment (MksmOrlov) [inline docs/scl-machine.tex:346]: Remove comment or uncomment comment?

File: problem-solver/cxx/solutionModule/manager/DeleteSolutionManager.cpp [hunk 3]
```diff
@@ -0,0 +1,84 @@
+#include "sc-agents-common/utils/IteratorUtils.hpp"
+
+#include "keynodes/SolutionKeynodes.hpp"
+
+#include "DeleteSolutionManager.hpp"
+
+namespace solution
+{
+DeleteSolutionManager::DeleteSolutionManager(ScMemoryContext * context)
```
[the remark this answers is not in the document]
Comment (MksmOrlov) [reply problem-solver/cxx/solutionModule/manager/DeleteSolutionManager.cpp:9]: And what?

File: problem-solver/cxx/solutionModule/manager/DeleteSolutionManager.hpp [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+namespace solution
+{
+class DeleteSolutionManager
+{
+public:
+  explicit DeleteSolutionManager(ScMemoryContext * context);
+
+  void deleteSolution(ScAddr const & solution) const;
+private:
+  ScMemoryContext * context;
```
[the remark this answers is not in the document]
Comment (MksmOrlov) [reply problem-solver/cxx/solutionModule/manager/DeleteSolutionManager.hpp:14]: And what if all will go to the window jumping hobby?

**Review by MksmOrlov — APPROVED** [review]

**Pull request merged**